lib/OpenQA/WebAPI/Plugin/ObsRsync/Controller.pm

```diff
@@ -0,0 +1,161 @@
+    $self->app->config->{obs_rsync}->{home};
+}
+
+sub index {
+    my $self   = shift;
+    my $folder = $self->param('folder');
+    return if $self->_check_and_render_error($folder);
+    my $files = Mojo::File->new($self->_home)->list({dir => 1})->map(sub { $_->basename })
+      ->grep(qr/^(?!test|WebAPIPlugin|__pycache__)/)->to_array;
+
+    my @files = grep -d Mojo::File->new($self->_home, $_), @$files;
+
+    $self->render('ObsRsync_index', folders => \@files);
+}
+
+sub folder {
+    my $self   = shift;
+    my $folder = $self->param('folder');
+    return if $self->_check_and_render_error($folder);
+
+    my $full        = $self->_home;
+    my $obs_project = $folder;
+    my $files       = Mojo::File->new($full, $obs_project, '.run_last')->list({dir => 1})->map(sub { $_->basename });
+
+    $self->render(
+        'ObsRsync_folder',
+        obs_project     => $obs_project,
+        lst_files       => $files->grep(qr/files_.*\.lst/)->to_array,
+        read_files_sh   => $files->grep(qr/read_files\.sh/)->join(),
+        rsync_commands  => $files->grep(qr/rsync_.*\.cmd/)->to_array,
+        rsync_sh        => $files->grep(qr/print_rsync.*\.sh/)->to_array,
+        openqa_commands => $files->grep(qr/openqa\.cmd/)->join(),
+        openqa_sh       => $files->grep(qr/print_openqa\.sh/)->join());
+}
+
+sub logs {
+    my $self   = shift;
+    my $folder = $self->param('folder');
+    return if $self->_check_and_render_error($folder);
+
+    my $full = Mojo::File->new($self->_home, $folder);
+    my $files
+      = $full->list({dir => 1, hidden => 1})->map(sub { $_->basename })->grep(qr/\.run_.*/)->sort(sub { $b cmp $a })
+      ->to_array;
+    $self->render('ObsRsync_logs', folder => $folder, full => $full->to_string, subfolders => $files);
+}
+
+sub logfiles {
+    my $self      = shift;
+    my $folder    = $self->param('folder');
+    my $subfolder = $self->param('subfolder');
+    return if $self->_check_and_render_error($folder, $subfolder);
+
+    my $full  = Mojo::File->new($self->_home, $folder, $subfolder);
+    my $files = $full->list({dir => 1})->map(sub { $_->basename })->sort->to_array;
+    $self->render(
+        'ObsRsync_logfiles',
+        folder    => $folder,
+        subfolder => $subfolder,
+        full      => $full->to_string,
+        files     => $files
+    );
+}
+
+sub download_file {
+    my $self      = shift;
+    my $folder    = $self->param('folder');
+    my $subfolder = $self->param('subfolder');
+    my $filename  = $self->param('filename');
+    return if $self->_check_and_render_error($folder, $subfolder, $filename);
+
+    my $full = $self->_home . '/' . $folder;
+    $full = $full . '/' . $subfolder if $subfolder;
+    $self->reply->file("$full/$filename");
+}
+
+sub run {
+    my $self   = shift;
+    my $folder = $self->param('folder');
+    return if $self->_check_and_render_error($folder);
+
+    my $cmd    = "bash '" . $self->_home . "/rsync.sh' '$folder' 2>&1";
+    my $out    = `$cmd`;
```

Comment:
This will need some hardening against injection attacks (even if it will be admin only in the future, why take risks with shell escaping). Modules like [IPC::System::Simple](https://metacpan.org/pod/IPC::System::Simple) tend to cover all bases.

Reply:
Let's do another PR for PUT method - I've removed it for now